README.md

```diff
@@ -84,6 +85,23 @@ Eg: `curl ... | NVM_DIR="path/to/nvm"`. Ensure that the `NVM_DIR` does not conta
 
 - The installer can use `git`, `curl`, or `wget` to download `nvm`, whichever is available.
 
+#### Installing in Docker
+
+To install `nvm` in Docker, you may need to prepare your image's environment in advance in order to use `node` & `npm`. Here's a winning recipe:
+
+```Dockerfile
+# set your target node version
+ENV NODE_VERSION 14.13.1
+# set NVM_DIR
+ENV NVM_DIR="$HOME/.nvm"
+# patch PATH
+ENV PATH $NVM_DIR/versions/node/v$NODE_VERSION/bin:$PATH
```

Comment:
this does not belong in docs; nobody should be manually editing the PATH, only sourcing `nvm.sh` and letting nvm do it.
```suggestion
```

it's also not necessary here since `nvm install` handles it in the next RUN statement.

Reply:
doesn't each run in docker happen in a new shell? iirc, mutating your env in a RUN doesn't persist in downstream runs

Reply:
Yes - but instead of hardcoding the PATH, is there a way the RUN that sources `nvm.sh` could export the resulting PATH for further RUNs to use?

Reply:
@ljharb @cdargine `ENV PATH="$NVM_DIR/versions/node/v$NODE_VERSION/bin:$PATH"` sholdnt this export the path to resulting run commands ?

Reply:
Ummm yeah I guess if thats not an option we can close this :+1:

Reply:
Would adding https://github.com/nvm-sh/nvm/pull/2318#discussion_r600659822 help docker users?

(note that there is a dockerfile already in this repo)

Reply:
> Would adding #2318 (comment) help docker users?

ya, certainly. it allows any execution context to use `nvm` and `node`, including (1) all container users and (2) in all downstream `RUN` stanzas, which otherwise isn't always feasible [without extra-work™](https://vsupalov.com/set-dynamic-environment-variable-during-docker-image-build/). this is particularly convenient when you want to use nvm/node in an image that isn't based on a node image, but needs node tools.

> note that there is a dockerfile already in this repo

Ya, this is Dockerfile is a great example of why such a patch as I sent originally is often needed. This repo's Dockerfile allows nvm usage in limited scope only. in the current state, to use the tool in docker:

- one must be executing as a user with a HOME, update that user's profile to source nvm content, and only execute the container as the associated user, OR
- source nvm.sh before every command

These constraints are not always convenient. What if I'm using node as part of a build only? What if I seek a global install? These are examples of when the current setup underperforms, and are common when I want to use nvm in base images that are primarily rooted in other language stacks.

Let's see these shortcomings in action:

case 1:

```sh
# fresh clone of this repo
❯ docker build -t nvm .
...<snip>
 => [24/24] WORKDIR /home/nvm/.nvm                                                                                           0.1s
 => exporting to image                                                                                                      10.7s
 => => exporting layers                                                                                                     10.7s
 => => writing image sha256:76cc557f01dedfd781a8fe3f871cd7c2db9ec9210f7f70d4de9e7e911047a64c                                 0.0s
 => => naming to docker.io/library/nvm

❯ docker run --rm -it --entrypoint sh nvm
$ node
sh: 1: node: not found
```

case 2:

Apply this patch:

```patch
diff --git a/Dockerfile b/Dockerfile
index 0d71868..050f848 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -105,6 +105,7 @@ RUN bash -c 'source $HOME/.nvm/nvm.sh   && \
     nvm install node                    && \
     npm install -g doctoc urchin eclint dockerfile_lint && \
     npm install --prefix "$HOME/.nvm/"'
+RUN node -v
 
 # Set WORKDIR to nvm directory
 WORKDIR /home/nvm/.nvm
```

Then, `docker build -t nvm .` yields `bin/bash: node: command not found`.



@ljharb , you already know a lot of this (it's just late and i'm kind of rapidly thinking out loud via typing 😄 ). But, point being, that patch as described makes nvm _more_ accessible in docker.

Because you recommend against meddling with the PATH, perhaps the correct advice for global nvm/node usage is to author an entry script, and put that on the path:

`/usr/bin/node`:
```bash
# pseudo-code
source $HOME/.nvm/nvm.sh > /dev/null
exec $(nvm node-path) $@
```

Reply:
@cdaringe is there a way we could update the Dockerfile in this repo (and let that serve as documentation) so that it did preserve the PATH changes nvm made, without needing to hardcode anything in the Dockerfile, such that your "case 2" there would work?

Reply:
@ljharb, i have been thinking about it, and I don't see a way forward otherwise, which is disappointing. ENV stanzas do not accept child commands, so we are left with only build args, the file system, and RUN blocks as sources of dynamic input, however only build arg inputs can be pulled into the image ENV definition from what i can tell.  or, better said, i cannot find a way to dynamically promote content from within a build step to the image ENV. if `ENV PATH=$PATH;$(nvm path)` worked, we would have a clean solve. it does not.

Reply:
I think I found a pretty neat way of installing nvm in a docker container by leveraging the bash environment variable `$BASH_ENV`, this way we don't even need to make the bash in the container into an interactive/login shell.

```Dockerfile
FROM ubuntu:22.10
SHELL ["/bin/bash", "-o", "pipefail", "-c"]

ENV BASH_ENV /root/.env
RUN touch $BASH_ENV
RUN echo '. "'"$BASH_ENV"'"' >> ~/.bashrc # only needed if executing interactively into the container

RUN apt-get update && apt-get install -y curl
RUN curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.2/install.sh | PROFILE="$BASH_ENV" bash

COPY .nvmrc .nvmrc
RUN nvm install
RUN node -v
```

This way nvm+node is easily available both later in the Dockerfile, and when executing into the container
```console
> docker build -t foo .
...
Step 9/10 : RUN nvm install
Found '//.nvmrc' with version <19.0.1>
...
Now using node v19.0.1 (npm v8.19.2)
Creating default alias: default -> 19.0.1 (-> v19.0.1 *)
...
Step 10/10 : RUN node -v
v19.0.1
> docker run --rm -it --entrypoint bash foo
root@4e5088bdde4d:/# tail -1 ~/.bashrc
. "/root/.env"
root@4e5088bdde4d:/# node -v
v19.0.1
```

Maybe it's neat enough to fit in the readme?

Reply:
Yeah I think it could be.

Sure!
https://github.com/blole/nvm/tree/docs/install-in-docker

I tried my hand at updating the Dockerfile in the repo as well, it seems to work like I'd expect, but I'm not sure what people use it for, so I can't really verify that.